### PR TITLE
 Reduce fingerprinting through timestamps in 'addr' messages.

### PR DESCRIPTION
**Pull request by sipa**
Suggested by @jonasnick.

### CODE REVIEW

**Comment by fanquake** [comment]
Travis failure
```
Cleaning up
terminate called after throwing an instance of 'boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::lock_error> >'
  what():  boost: mutex lock failed in pthread_mutex_lock: Invalid argument
terminate called after throwing an instance of 'boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::lock_error> >'
  what():  boost: mutex lock failed in pthread_mutex_lock: Invalid argument
*** glibc detected *** /home/travis/build/bitcoin/bitcoin/bitcoin-x86_64-unknown-linux-gnu/src/bitcoind: malloc(): memory corruption: 0x00007fd8131c3ab1 ***
```

**Comment by Diapolo** [comment]
@laanwj Isn't 0.10.0 released ;)? Why did you tag this 0.10.0 then?

**Comment by laanwj** [comment]
@diapola Don't be so pedantic, I'm using it as marker for the next version on the 0.10 branch.

**Comment by Diapolo** [comment]
@laanwj Just asking, no offense :).

**Comment by laanwj** [comment]
@diapolo okay :)

@sipa @fanquake Tried a Travis respin, still same problem on same configuration.

**Comment by laanwj** [comment]
I'm also getting these errors when running the RPC tests locally with this patch:
```
bitcoind: /usr/include/boost/thread/pthread/recursive_mutex.hpp:110: void boost::recursive_mutex::lock(): Assertion `!pthread_mutex_lock(&m)' failed.
bitcoind: /usr/include/boost/thread/pthread/recursive_mutex.hpp:110: void boost::recursive_mutex::lock(): Assertion `!pthread_mutex_lock(&m)' failed.
```
This appears to be a problem with destruction order. By moving `addrman.Connected(addr);` to the CNode destructor, addrman is called when each node is deleted. This may cause a use-after-free of a mutex?

**Comment by jgarzik** [comment]
concept ACK

**Comment by sipa** [comment]
Changed approach.

**Comment by jonasnick** [comment]
To give a bit of context, without this PR it is possible to learn all outbound peers of a node by repeatedly sending 'getaddr' messages. See the histogram for the timestamps of a full node and a client client.

**Comment by laanwj** [comment]
utACK

File: src/main.cpp [hunk 1]
```diff
@@ -3628,6 +3638,12 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
     else if (strCommand == "verack")
     {
         pfrom->SetRecvVersion(min(pfrom->nVersion, PROTOCOL_VERSION));
+
+        // Update the last seen time for this node's address
```
Comment (laanwj) [inline src/main.cpp]: This comment is no longer correct, should probably be `Update the connected status for this node`

**Comment by jonasnick** [comment]
untested ack

File: src/main.cpp (hunk 1 above)
Comment (sipa) [inline src/main.cpp]: Done.

**Comment by laanwj** [comment]
Cherry-picked to 0.10 as ca301bf

**Pull request merged**